### PR TITLE
Draft pull request : Devon kafka implementation

### PR DESCRIPTION
**Pull request by ravicm2**
This is a draft pull request for code review of Kafka producer and consumer implementation.

### CODE REVIEW

**Comment by ravicm2** [comment]
i will do git squash when the code merges.

File: modules/kafka/src/main/java/com/devonfw/module/common/config/KafkaConsumerProperties.java [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+    this.autoOffsetReset = autoOffsetReset;
+    this.fetchMinBytes = fetchMinBytes;
+    this.fetchMaxBytes = fetchMaxBytes;
+    this.fetchMaxWaitMs = fetchMaxWaitMs;
+    this.maxPartitionFetchBytes = maxPartitionFetchBytes;
+    this.isolationLevel = isolationLevel;
+  }
+
+  /**
+   * @return bootstrapservers
```
[the remark this answers is not in the document]
Comment (hohwille) [reply modules/kafka/src/main/java/com/devonfw/module/common/config/KafkaConsumerProperties.java]: https://github.com/devonfw/devon4j/blob/develop/documentation/guide-configuration-mapping.asciidoc

File: modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  /**
+   * This method is used to throw throwIllegalArgumentException when payloads are null.
+   *
+   */
+  private void throwIllegalArgumentException(String message) {
+
+    throw new IllegalArgumentException(message);
+  }
```
Comment (hohwille) [inline modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java]: I do not see any value in creating a method for this.

File: modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+  /**
+   * This method is used to set consumer properties.
+   *
+   * @param kafkaConsumerProperties
+   * @return
+   */
+  private Properties setConsumerProperties(KafkaConsumerProperties kafkaConsumerProperties) {
```
Comment (hohwille) [inline modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java]: so if this method converts `KafkaConsumerProperties` to regular `Properties`:
* why is the method called `setConsumerProperties`? It is not a setter!
* why not instead add an `asProperties()` or `toProperties()` method directly `KafkaConsumerProperties`?

File: modules/kafka/src/main/java/com/devonfw/module/producer/KafkaMessageSender.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+  }
+
+  /**
+   * This method is used to prepare Properties payload.
+   *
+   * @param producerProperties
+   * @return
+   */
+  private Properties setProducerPropeties(KafkaProducerProperties producerProperties) {
```
Comment (hohwille) [inline modules/kafka/src/main/java/com/devonfw/module/producer/KafkaMessageSender.java]: same here...

File: modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ *
+ */
+public class ProducerPayload {
+
+  private Object message;
+
+  /**
+   * Topic name. If topic name is not provided then default name will be assigned.
+   */
+  private String topic;
```
Comment (hohwille) [inline modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java]: Nice that in this case you actually add JavaDoc that is helpful. However, this should go to the getter and not stay here where it never gets into the generated JavaDocs and nobody can ever see it in IDE auto-completion.

File: modules/kafka/src/test/java/com/devonfw/module/consumer/KafkaMessageReceiverTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  private EmbeddedKafkaBroker embeddedKafkaBroker;
+
+  private KafkaMessageReceiver KafkaMessageReceiver;
+
+  /**
+   *
+   */
+  @BeforeEach
+  public void setup() {
```
Comment (hohwille) [inline modules/kafka/src/test/java/com/devonfw/module/consumer/KafkaMessageReceiverTest.java]: override `doSetup()` instead.

File: modules/kafka/src/test/java/com/devonfw/module/consumer/KafkaMessageReceiverTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    this.KafkaMessageReceiver = new KafkaMessageReceiver();
+
+    this.embeddedKafkaBroker = new EmbeddedKafkaBroker(1, true, TOPIC);
+  }
+
+  /**
+   *
+   */
```
Comment (hohwille) [inline modules/kafka/src/test/java/com/devonfw/module/consumer/KafkaMessageReceiverTest.java]: You do not need to add JavaDoc here. But if you do then also fill it. Otherwise this is just a waste of lines of code.

File: modules/pom.xml [hunk 8]
```diff
@@ -1,120 +1,134 @@
...
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>com.devonfw.java.dev</groupId>
+		<artifactId>devon4j</artifactId>
+		<version>dev-SNAPSHOT</version>
+	</parent>
+	<artifactId>devon4j-modules</artifactId>
+	<packaging>pom</packaging>
+	<name>${project.artifactId}</name>
+	<description>Reusable modules of devon4j.</description>
```
Comment (hohwille) [inline modules/pom.xml]: could you follow devonfw code style? Your IDE should be preconfigured with indendation by 2 spaces. Just hit [Ctrl][Shift][f]. This will help use to prevent diff wars in git. I am struggling often when doing reviews because of that when there are tons of lines in the diff that actually did not change but people where just using different formatters. I thought this was solved with `devonfw-ide`. Seems we still have to invest more on this...

**Comment by hohwille** [comment]
I know this is still in draft state, but do you also plan to add a `guide-kafka.asciidoc` to the documentation?

File: modules/pom.xml (hunk 8 above)
Comment (ravicm2) [reply modules/pom.xml]: i did that CTRL+SHIFT+F but still it is same . let me try one more time

**Comment by ravicm2** [comment]
> I know this is still in draft state, but do you also plan to add a `guide-kafka.asciidoc` to the documentation?

yes jorg after the complete development i will be adding it .

File: modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java (hunk 2 above)
Comment (ravicm2) [reply modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java]: as the exception is throw multiple times i thought of keeping it as a method

File: modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
+package com.devonfw.module.producer;
+
+/**
+ * @author ravicm
+ *
+ */
+public class ProducerPayload {
```
Comment (maihacke) [inline modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java]: I don't think it is a good idea to "bundle" topic name and payload.  Why is that necessary?

File: modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * The constructor.
+   *
+   * @param message
+   * @param topic
+   * @param key
+   * @param partition
+   * @param timestamp
+   */
+  public ProducerPayload(Object message, String topic, String key, Integer partition, Long timestamp) {
```
Comment (maihacke) [inline modules/kafka/src/main/java/com/devonfw/module/producer/ProducerPayload.java]: Why do we set the partition number here explicitly? Normally that is done by kafka/spring-kafka

File: modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+  /**
+   * This method is used to receiver message from the kafka broker based on topic names and group names. If groupName is
+   * not provided all the messages belongs to the topic names will be returned
+   *
+   * @param topicNames
+   * @param kafkaConsumerProperties
+   * @return
+   */
+  public ConsumerRecords<Object, Object> receiveMessage(List<String> topicNames,
```
[the remark this answers is not in the document]
Comment (ravicm2) [reply modules/kafka/src/main/java/com/devonfw/module/consumer/KafkaMessageReceiver.java]: @maihacke So, we dont need to write consumer class?
